File: src/main/java/seremis/geninfusion/soul/traits/TraitNavigate.java
package seremis.geninfusion.soul.traits;

import net.minecraft.entity.Entity;
import net.minecraft.init.Blocks;
import net.minecraft.util.MathHelper;
import seremis.geninfusion.api.soul.IEntitySoulCustom;
import seremis.geninfusion.api.soul.SoulHelper;
import seremis.geninfusion.api.soul.lib.Genes;

public class TraitNavigate extends Trait {

    @Override
    public void updateWanderPath(IEntitySoulCustom entity) {
        double posX = entity.getDouble("posX");
        double posY = entity.getDouble("posY");
        double posZ = entity.getDouble("posZ");

        entity.getWorld().theProfiler.startSection("stroll");
        boolean flag = false;
        int i = -1;
        int j = -1;
        int k = -1;
        float f = -99999.0F;

        for(int l = 0; l < 10; ++l) {
            int i1 = MathHelper.floor_double(posX + (double) entity.getRandom().nextInt(13) - 6.0D);
            int j1 = MathHelper.floor_double(posY + (double) entity.getRandom().nextInt(7) - 3.0D);
            int k1 = MathHelper.floor_double(posZ + (double) entity.getRandom().nextInt(13) - 6.0D);
            float f1 = entity.getBlockPathWeight(i1, j1, k1);

            if(f1 > f) {
                f = f1;
                i = i1;
                j = j1;
                k = k1;
                flag = true;
            }
        }

        if(flag) {
            entity.setObject("pathToEntity", entity.getWorld().getEntityPathToXYZ((Entity) entity, i, j, k, 10.0F, true, false, false, true));
        }

        entity.getWorld().theProfiler.endSection();
    }

    @Override
    public float getBlockPathWeight(IEntitySoulCustom entity, int x, int y, int z) {
        boolean burnsInDaylight = SoulHelper.geneRegistry().getValueFromAllele(entity, Genes.GENE_BURNS_IN_DAYLIGHT);
        boolean eatsGrass = SoulHelper.geneRegistry().getValueFromAllele(entity, Genes.GENE_AI_EAT_GRASS);

        if(burnsInDaylight) {
            return 0.5F - entity.getWorld().getLightBrightness(x, y, z);
        }
        if(eatsGrass) {
            return entity.getWorld().getBlock(x, y - 1, z) == Blocks.grass ? 10.0F : entity.getWorld().getLightBrightness(x, y, z) - 0.5F;
        }
        //TODO silverfish in stone block
        return 0.0F;
    }

    @Override
    public Entity findPlayerToAttack(IEntitySoulCustom entity) {
        return entity.getWorld().getClosestPlayerToEntity((Entity) entity, 50);
    }
}

File: src/main/java/seremis/geninfusion/soul/traits/TraitNavigate.scala
package seremis.geninfusion.soul.traits

import net.minecraft.entity.Entity
import net.minecraft.init.Blocks
import net.minecraft.util.MathHelper
import seremis.geninfusion.api.soul.lib.Genes
import seremis.geninfusion.api.soul.lib.VariableLib._
import seremis.geninfusion.api.soul.{IEntitySoulCustom, SoulHelper}

class TraitNavigate extends Trait {

    override def updateWanderPath(entity: IEntitySoulCustom) {
        val posX = entity.getDouble(ENTITY_POS_X)
        val posY = entity.getDouble(ENTITY_POS_Y)
        val posZ = entity.getDouble(ENTITY_POS_Z)

        entity.getWorld.theProfiler.startSection("stroll")

        var pathFound = false
        var destX = -1
        var destY = -1
        var destZ = -1
        var blockWeight = -99999.0F

        for(l <- 0 until 10) {
            val randX = MathHelper.floor_double(posX + entity.getRandom.nextInt(13).toDouble - 6.0D)
            val randY = MathHelper.floor_double(posY + entity.getRandom.nextInt(7).toDouble - 3.0D)
            val randZ = MathHelper.floor_double(posZ + entity.getRandom.nextInt(13).toDouble - 6.0D)
            val randBlockWeight = entity.getBlockPathWeight(randX, randY, randZ)

            if(randBlockWeight > blockWeight) {
                blockWeight = randBlockWeight
                destX = randX
                destY = randY
                destZ = randZ
                pathFound = true
            }
        }

        if(pathFound) {
            entity.setObject(ENTITY_PATH_TO_ENTITY, entity.getWorld.getEntityPathToXYZ(entity.asInstanceOf[Entity], destX, destY, destZ, 10.0F, true, false, false, true))
        }
        entity.getWorld.theProfiler.endSection()
    }

    override def getBlockPathWeight(entity: IEntitySoulCustom, x: Int, y: Int, z: Int): Float = {
        val burnsInDaylight = SoulHelper.geneRegistry.getValueFromAllele[Boolean](entity, Genes.GENE_BURNS_IN_DAYLIGHT)
        val eatsGrass = SoulHelper.geneRegistry.getValueFromAllele[Boolean](entity, Genes.GENE_AI_EAT_GRASS)

        if(burnsInDaylight) {
            return 0.5F - entity.getWorld.getLightBrightness(x, y, z)
        }

        if(eatsGrass) {
            return if(entity.getWorld.getBlock(x, y - 1, z) == Blocks.grass) 10.0F else entity.getWorld.getLightBrightness(x, y, z) - 0.5F
        }
        0.0F
    }

    override def findPlayerToAttack(entity: IEntitySoulCustom): Entity = {
        entity.getWorld.getClosestPlayerToEntity(entity.asInstanceOf[Entity], 50)
    }
}